server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -80,12 +80,25 @@ protected TransportWriteAction(Settings settings, String actionName, TransportSe
 
     @Override
     protected Releasable checkOperationLimits(Request request) {
-        return writeMemoryLimits.markCoordinatingOperationStarted(primaryOperationSize(request));
+        assert supportsRerouteAction() : "checkOperationLimits should be be called if reroute not supported by action";
+        return writeMemoryLimits.markWriteOperationStarted(primaryOperationSize(request));
     }
 
     @Override
     protected Releasable checkPrimaryLimits(Request request) {
-        return writeMemoryLimits.markPrimaryOperationStarted(primaryOperationSize(request));
+        if (supportsRerouteAction()) {
+            return () -> {};
```

Comment:
One scenario which is not supported by this I think is if we have a relocating primary where the old primary has done a relocation handoff with the new primary, and is directly sending a primary action from old primary to new primary (see `primaryShardReference.isRelocated()` in `TransportReplicationAction.AsyncPrimaryAction.runWithPrimaryShardReference`).

The new primary would then not account for the bytes. This should only be a for a short time, nevertheless I'm wondering if we can find a solution for this.